functions/index.js

```diff
@@ -1,8 +1,122 @@
+// The Cloud Functions for Firebase SDK to create Cloud Functions and setup triggers.
 const functions = require('firebase-functions');
 
+// The Firebase Admin SDK to access the Firebase Realtime Database.
+const admin = require('firebase-admin');
+admin.initializeApp();
+
 // // Create and Deploy Your First Cloud Functions
 // // https://firebase.google.com/docs/functions/write-firebase-functions
 //
 // exports.helloWorld = functions.https.onRequest((request, response) => {
 //  response.send("Hello from Firebase!");
 // });
+
+// Take the text parameter passed to this HTTP endpoint and insert it into the
+// Realtime Database under the path /messages/:pushId/original
+exports.addMessage = functions.https.onRequest((req, res) => {
+  // Grab the text parameter.
+  const original = req.query.text;
+  // Push the new message into the Realtime Database using the Firebase Admin SDK.
+  return admin.firestore().collection('messages').add({original: original}).then((writeResult) => {
+    // Send back a message that we've succesfully written the message
+    return res.json({result: `Message with ID: ${writeResult.id} added.`});
+  });
+  // [END adminSdkAdd]
+});
+// [END addMessage]
+
+exports.clearGarbagePeople = functions.https.onRequest((req, res) => {
+  
+  return admin.firestore().collection('persons').get().then((queryResult) => {
+    var numBadPersons = 0;
+    var batch = admin.firestore().batch();
+    queryResult.forEach((doc)=> {
+      // everybody needs a timestamp
+      if (typeof doc.data().timestamp === 'undefined') {
+        batch.delete(doc.ref);
+        numBadPersons += 1;
+        if (numBadPersons > 450) {
+          batch.commit();
+          return res.json({result: `more than 450 bad people, rerun function`});
+        }
+      }
+    });
+    batch.commit();
+    return res.json({result: `Successfully deleted ${numBadPersons} persons`});
+  })
+  .catch((error) => {
+    return res.json({result:`${error}`})
+  });
+});
+
+exports.deduplicatePersons = functions.https.onRequest((req, res) => {
+  
+  return admin.firestore().collection('persons').get().then((queryResult) => {
+    var ben = {name:'ben'};
+
+    var persons = new Map();
+    var duplicatePersons = new Map();
+    queryResult.forEach((result)=> {
+      var person = result.data();
+      //todo make this get rid of capitalization etc
+      person.id = result.id;
+      if (typeof persons.get(person.name) === 'undefined') {
+        persons.set(person.name, person);
```

Comment:
if person.name is undefined, you are setting person.name to person? Am I reading this right? I know it's kind of late as I am writing this message, so pardon me.